polygeist/lib/polygeist/Passes/ConvertToLLVMABI.cpp

```diff
@@ -178,6 +178,74 @@ class SYCLConstructorLowering
   }
 };
 
+static LogicalResult convertMethod(sycl::SYCLMethodOpInterface method,
+                                   PatternRewriter &rewriter) {
+  LLVM_DEBUG(llvm::dbgs() << "ConvertToLLVMABIPass: SYCLMethodOpLowering: ";
+             method.dump(); llvm::dbgs() << "\n");
+
+  SmallVector<mlir::Value> Args(method->getOperands());
+
+  const auto BaseType = method.getBaseType().cast<MemRefType>();
+  const auto ThisType = Args[0].getType().cast<MemRefType>();
+  if (BaseType.getElementType() != ThisType.getElementType()) {
+    const auto TargetType =
+        MemRefType::get(BaseType.getShape(), BaseType.getElementType(), {},
+                        ThisType.getMemorySpaceAsInt());
```

Comment:
Here you want to cast `ThisType` to the `BaseType` when the element type of both MemRefs do not match. 
Do you expect the shape and the address space of the 2 MemRef to always match? I am wondering whether at line 190 we should check for any mismatch (not just different element type) and always cast `ThisType` to `BaseType` (which potentially involves changing the shape, and/or the lement type and/or the adddress space of the `ThisType`)

Reply:
As we're removing the `SYCLCastOp` (and `addressspacecasts`), we only expect differences in address space and element type. I'll add an address space cast here.
I'll add an assertion to check that the shape is correct anyway, that's a good idea.